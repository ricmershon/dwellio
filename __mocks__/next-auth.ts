export const getServerSession = jest.fn()
export const signIn = jest.fn()
export const signOut = jest.fn()

export const useSession = jest.fn(() => ({
    data: null,
    status: 'unauthenticated'
}))

export const SessionProvider = ({ children }: { children: React.ReactNode }) => children

const nextAuth = {
    getServerSession,
    signIn,
    signOut,
    useSession,
    SessionProvider
}

export default nextAuth