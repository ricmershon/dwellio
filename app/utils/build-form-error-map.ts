import { ZodError } from "zod";

export const buildFormErrorMap = (issues: ZodError['issues']) => {
    console.log(issues);
    const errorMap: Record<string, string[]> = {};

    for (const issue of issues) {
        const path = issue.path.join('.');
        if (!errorMap[path]) {
            errorMap[path] = [];
        }
        errorMap[path].push(issue.message);
    }

    return structureErrors(errorMap);
};

type ErrorMap = Record<string, string[]>;

interface StructuredErrorMap {
    location?: Record<string, string[]>;
    rates?: Record<string, string[]>;
    sellerInfo?: Record<string, string[]>;
    [key: string]: Record<string, string[]> | string[] | undefined;
}

function structureErrors(flatErrors: ErrorMap): StructuredErrorMap {
    const structured: StructuredErrorMap = {};

    for (const [key, messages] of Object.entries(flatErrors)) {
        if (key.startsWith("location.")) {
            structured.location = structured.location || {};
            const field = key.split(".")[1];
            structured.location[field] = messages;
        } else if (key.startsWith("rates.")) {
            structured.rates = structured.rates || {};
            const field = key.split(".")[1];
            structured.rates[field] = messages;
        } else if (key.startsWith("sellerInfo.")) {
            structured.sellerInfo = structured.sellerInfo || {};
            const field = key.split(".")[1];
            structured.sellerInfo[field] = messages;
        } else {
            structured[key] = messages;
        }
    }

    return structured;
};