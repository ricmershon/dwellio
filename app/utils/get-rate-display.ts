import { Rates } from "@/app/types/definitions";

export const getRateDisplay = (rates: Rates) => {
    if (rates.monthly) {
        return `$${rates.monthly.toLocaleString()}/mo`;
    } else if (rates.weekly) {
        return `$${rates.weekly.toLocaleString()}/wk`;
    } else if (rates.nightly) {
        return `$${rates.nightly.toLocaleString()}/night`;
    }
};