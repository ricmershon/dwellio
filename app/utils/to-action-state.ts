import { ActionState } from "@/app/lib/definitions";

export const toActionState = (
    message: string,
    status: "SUCCESS" | "ERROR",
    isBookmarked?: boolean
): ActionState => ({ message, status, isBookmarked });
