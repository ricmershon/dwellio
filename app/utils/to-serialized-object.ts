import { MessageDocument, PropertyDocument } from "@/app/models";

// TODO: need a cleaner solution
export const toSerializedOjbect = (object: PropertyDocument | PropertyDocument[] | MessageDocument) => (
    JSON.parse(JSON.stringify(object))
);