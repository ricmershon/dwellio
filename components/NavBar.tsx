'use client';

import { useState } from "react";
import Link from "next/link";
import Image from "next/image";
import { usePathname } from "next/navigation";
import { BellAlertIcon } from "@heroicons/react/24/outline";
import { HomeIcon } from "@heroicons/react/24/solid";
import { FaGoogle } from 'react-icons/fa'

import profileDefault from '@/assets/images/profile.png';

const NavBar = () => {
    const [isMobileMenuOpen, setIsMobileMenuOpen] = useState(false);
    const [isProfileDropdownOpen, setIsProfileDropdownOpen] = useState(false);
    const [isLoggedIn, setIsLoggedIn] = useState(false);

    const pathname = usePathname();

    return (
        <nav className='bg-blue-700 border-b border-blue-500'>
            <div className='mx-auto max-w-7xl px-2 sm:px-6 lg:px-8'>
                <div className='relative flex h-18 items-center justify-between'>

                    {/* Mobile menu button */}
                    <div className='absolute inset-y-0 left-0 flex items-center md:hidden'>

                        <button
                            type='button'
                            id='mobile-dropdown-button'
                            className='relative inline-flex items-center justify-center rounded-md p-2 text-gray-400 hover:bg-gray-700 hover:text-white focus:outline-none focus:ring-2 focus:ring-inset focus:ring-white'
                            aria-controls='mobile-menu'
                            aria-expanded='false'
                            onClick={() => setIsMobileMenuOpen((prevState) => !prevState)}
                        >
                            <span className='absolute -inset-0.5'></span>
                            <span className='sr-only'>Open main menu</span>
                            <svg
                                className='block h-6 w-6'
                                fill='none'
                                viewBox='0 0 24 24'
                                strokeWidth='1.5'
                                stroke='currentColor'
                                aria-hidden='true'
                            >
                                <path
                                    strokeLinecap='round'
                                    strokeLinejoin='round'
                                    d='M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5'
                                />
                            </svg>
                        </button>
                    </div>

                    {/* Menu left side */}
                    <div className='flex flex-1 items-center justify-center md:justify-start'>

                        {/* Logo */}
                        <Link
                            className='flex flex-shrink-0 items-center' href='/' >
                            <HomeIcon className='h-10 w-auto text-blue-700 rounded-full p-[4px] bg-white' />
                            <span
                                className="hidden md:block text-xl text-white ml-2"
                            >
                                dwellio
                            </span>
                        </Link>

                        {/* Desktop menu left side */}
                        <div className="hidden md:ml-8 md:block">
                            <div className="flex space-x-5">
                                <Link
                                    href='/'
                                    className={`${pathname === '/'
                                        && 'border-b border-white'
                                    } text-white hover:border-b hover:border-white`}
                                >
                                    Home
                                </Link>
                                <Link
                                    href='/properties'
                                    className={`${pathname === '/properties'
                                        && 'border-b border-white'
                                    } text-white hover:border-b hover:border-white`}
                                >
                                    Properties
                                </Link>
                                {isLoggedIn && (
                                    <Link
                                        href='/properties/add'
                                        className={`${pathname === '/properties/add'
                                            && 'border-b border-white'
                                        } text-white hover:border-b hover:border-white`}
                                    >
                                        Add Property
                                    </Link>                        
                                )}
                            </div>
                        </div>
                    </div>

                    {/* Right side menus */}
                    {isLoggedIn ? (
                        // Logged in
                        <div className="absolute inset-y-0 right-0 flex items-center pr-2 md:static md:inset-auto md:ml-6 md:pr-0">
                            <Link className='relative group' href='/messages'>
                                <BellAlertIcon className='relative h-8 w-8 rounded-full bg-gray-800 p-1 text-gray-400 hover:text-white focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2 focus:ring-offset-gray-800'/>
                                <span className="absolute top-0 right-0 inline-flex items-center justify-center h-[18px] w-[18px] text-[12px] leading-none text-white transform translate-x-1/2 -translate-y-1/2 bg-red-500 rounded-full">
                                    1
                                </span>
                            </Link>
    
                            {/* Profile dropdown button */}
                            <div className="relative ml-5">
                                <div>
                                    <button
                                        type="button"
                                        className="relative flex rounded-full bg-gray-800 text-sm focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2 focus:ring-offset-gray-800"
                                        id="user-menu-button"
                                        aria-expanded="false"
                                        aria-haspopup="true"
                                        onClick={() => setIsProfileDropdownOpen((prevState) => !prevState)}
                                    >
                                    <span className="absolute -inset-1.5"></span>
                                    <span className="sr-only">Open user menu</span>
                                    <Image
                                        className="h-8 w-8 rounded-full"
                                        height={32}
                                        width={32}
                                        src={profileDefault}
                                        alt=""
                                    />
                                    </button>
                                </div>
    
                                {/* Profile dropdown menu */}
                                {isProfileDropdownOpen && (
                                    <div
                                        id="user-menu"
                                        className="absolute right-0 z-10 mt-2 w-48 origin-top-right rounded-md bg-white py-1 shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none"
                                        role="menu"
                                        aria-orientation="vertical"
                                        aria-labelledby="user-menu-button"
                                        tabIndex={-1}
                                    >
                                        <Link
                                            href="/profile"
                                            className="block px-4 py-2 text-sm text-gray-700"
                                            role="menuitem"
                                            tabIndex={-1}
                                            id="user-menu-item-0"
                                        >
                                            Your Profile
                                        </Link>
                                        <Link
                                            href="/properties/saved"
                                            className="block px-4 py-2 text-sm text-gray-700"
                                            role="menuitem"
                                            tabIndex={-1}
                                            id="user-menu-item-2"
                                        >
                                            Saved Properties
                                        </Link>
                                        <button
                                            className="block px-4 py-2 text-sm text-gray-700"
                                            role="menuitem"
                                            tabIndex={-1}
                                            id="user-menu-item-2"
                                        >
                                            Sign Out
                                        </button>
                                    </div>
                                )}
                            </div>
                        </div>
                    ) : (
                        // Logged out
                        <div className="hidden md:block md:ml-6">
                            <div className="flex items-center">
                                <button className="flex items-center text-white bg-gray-700 hover:bg-gray-900 hover:text-white rounded-md px-3 py-2 my-5">
                                    <FaGoogle className='text-white mr-2' />
                                    <span>Login or Register</span>
                                </button>
                            </div>
                        </div>
                    )}
                </div>
            </div>

            {/* Mobile menu, items shown or hidden based on menu state */}
            {isMobileMenuOpen && (
                <div id='mobile-menu'>
                    <div className="space-y-1 px-2 pb-3 pt-2">
                        <Link
                            href='/'
                            className={`${pathname === '/' && 'bg-black'} text-white block rounded-md px-3 py-2 text-base font-medium`}
                        >
                            Home
                        </Link>
                        <Link
                            href='/properties'
                            className={`${pathname === '/properties' && 'bg-black'} text-white block rounded-md px-3 py-2 text-base font-medium`}
                        >
                            Properties
                        </Link>
                        {isLoggedIn ? (
                            <Link
                                href='/properties/add'
                                className={`${pathname === '/properties/add' && 'bg-black'} text-white block rounded-md px-3 py-2 text-base font-medium`}
                            >
                                Add Property
                            </Link>
                        ) : (
                            <button className="flex items-center text-white bg-gray-700 hover:bg-gray-900 hover:text-white rounded-md px-3 py-2 my-2">
                                <FaGoogle className='text-white mr-2' />
                                <span>Login or Register</span>
                            </button>
                        )}
                    </div>
                </div>
            )}
        </nav>
    );
}
 
export default NavBar;