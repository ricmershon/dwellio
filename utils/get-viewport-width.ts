import { cookies } from "next/headers";

import { VIEWPORT_WIDTH_COOKIE_NAME } from "@/types";

/**
 * Retrieves viewport width from browser cookies
 * @returns Viewport width in pixels, or 0 if not available or invalid
 */
export const getViewportWidth = async () => {
    try {
        // Validate VIEWPORT_WIDTH_COOKIE_NAME is defined
        if (!VIEWPORT_WIDTH_COOKIE_NAME) {
            console.warn('getViewportWidth: VIEWPORT_WIDTH_COOKIE_NAME is not defined');
            return 0;
        }
        
        const cookieStore = await cookies();
        
        // Validate cookieStore was retrieved successfully
        if (!cookieStore) {
            console.warn('getViewportWidth: cookies() returned null or undefined');
            return 0;
        }
        
        const viewportCookie = cookieStore.get(VIEWPORT_WIDTH_COOKIE_NAME)?.value;
        
        // Check if cookie exists and has a value
        if (!viewportCookie) {
            // This is normal - no warning needed for missing viewport cookie
            return 0;
        }
        
        // Validate cookie value is a string
        if (typeof viewportCookie !== 'string') {
            console.warn('getViewportWidth: viewport cookie value is not a string');
            return 0;
        }
        
        // Convert to number with validation
        const width = Number(viewportCookie);
        
        // Validate the conversion was successful
        if (isNaN(width)) {
            console.warn('getViewportWidth: viewport cookie value is not a valid number:', viewportCookie);
            return 0;
        }
        
        // Validate width is a reasonable viewport size
        if (width < 0) {
            console.warn('getViewportWidth: negative viewport width detected:', width);
            return 0;
        }
        
        if (width > 10000) {
            console.warn('getViewportWidth: suspiciously large viewport width detected:', width);
            // Don't return 0 here as ultra-wide monitors exist, but log for monitoring
        }
        
        return width;
    } catch (error) {
        // Only log unexpected errors in development or non-build scenarios
        if (process.env.NODE_ENV === 'development' || 
            (error instanceof Error && !error.message.includes('DYNAMIC_SERVER_USAGE'))) {
            console.error('getViewportWidth: error retrieving viewport width:', error);
        }
        return 0;
    }
}