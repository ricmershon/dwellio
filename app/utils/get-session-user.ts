import { getServerSession } from "next-auth/next";
import { authConfig } from "@/app/config/auth-config";

export const getSessionUser = async () => {
    const session = await getServerSession(authConfig);
    if (!session || !session.user) {
        return null;
    }
    return session.user
}