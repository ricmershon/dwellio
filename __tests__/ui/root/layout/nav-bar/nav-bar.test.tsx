import React from 'react';
import { render, screen } from '@testing-library/react';
import { createNextNavigationMock } from '@/__tests__/test-utils';
import { useSession } from 'next-auth/react';
import { usePathname } from 'next/navigation';

import NavBar from '@/ui/root/layout/nav-bar/nav-bar';
import { getViewportWidth } from '@/utils/get-viewport-width';

// Mock all the sub-components and their dependencies
jest.mock('next-auth/react', () => ({
    useSession: jest.fn(),
}));

jest.mock('next/navigation', () => createNextNavigationMock());

jest.mock('@/utils/get-viewport-width', () => ({
    getViewportWidth: jest.fn(),
}));

jest.mock('@/ui/root/layout/nav-bar/nav-bar-left', () => ({
    __esModule: true,
    default: () => <div data-testid="nav-bar-left">NavBarLeft</div>,
}));

// Simple mocks that don't try to use hooks in jest.mock
jest.mock('@/ui/root/layout/nav-bar/nav-bar-desktop-middle', () => ({
    __esModule: true,
    default: () => <div data-testid="nav-bar-desktop-middle">NavBarDesktopMiddle</div>,
}));

jest.mock('@/ui/root/layout/nav-bar/nav-bar-desktop-right', () => ({
    __esModule: true,
    default: ({ viewportWidth }: { viewportWidth: number }) => (
        <div data-testid="nav-bar-desktop-right" data-viewport-width={viewportWidth}>
            NavBarDesktopRight
        </div>
    ),
}));

jest.mock('@/ui/root/layout/nav-bar/nav-bar-right', () => ({
    __esModule: true,
    default: () => <div data-testid="nav-bar-right">NavBarRight</div>,
}));

jest.mock('@/context/global-context', () => ({
    useGlobalContext: () => ({
        unreadCount: 3,
    }),
}));

jest.mock('@/hooks/use-click-outside', () => ({
    useClickOutside: jest.fn(),
}));

jest.mock('@/ui/auth/login-buttons', () => ({
    __esModule: true,
    default: ({ buttonClassName, text, icon }: {
        buttonClassName?: string;
        text?: string;
        icon?: React.ReactNode;
    }) => (
        <div data-testid="login-buttons-component">
            <button className={buttonClassName}>
                {icon}
                {text}
            </button>
        </div>
    ),
}));

jest.mock('@/ui/auth/logout-button', () => ({
    __esModule: true,
    default: ({ setIsMenuOpen, id }: {
        setIsMenuOpen?: (value: boolean) => void;
        id?: string;
    }) => (
        <button 
            data-testid="logout-button" 
            id={id}
            onClick={() => setIsMenuOpen?.(false)}
        >
            Logout
        </button>
    ),
}));

describe('NavBar', () => {
    const mockUseSession = useSession as jest.MockedFunction<typeof useSession>;
    const mockUsePathname = usePathname as jest.MockedFunction<typeof usePathname>;
    const mockGetViewportWidth = getViewportWidth as jest.MockedFunction<typeof getViewportWidth>;
    
    const mockUser = {
        id: '1',
        name: 'John Doe',
        email: 'john@example.com',
        image: 'https://example.com/profile.jpg',
    };
    
    const mockSession = {
        user: mockUser,
        expires: new Date(Date.now() + 24 * 60 * 60 * 1000).toISOString(),
    };

    beforeEach(() => {
        jest.clearAllMocks();
        mockUsePathname.mockReturnValue('/');
        mockGetViewportWidth.mockResolvedValue(1024);
    });

    describe('Component Structure', () => {
        it('should render the main navigation structure', async () => {
            mockUseSession.mockReturnValue({ data: null, status: 'unauthenticated' } as any);
            
            render(await NavBar());
            
            expect(screen.getByRole('navigation')).toBeInTheDocument();
            expect(screen.getByTestId('nav-bar-left')).toBeInTheDocument();
            expect(screen.getByTestId('nav-bar-desktop-middle')).toBeInTheDocument();
            expect(screen.getByTestId('nav-bar-desktop-right')).toBeInTheDocument();
            expect(screen.getByTestId('nav-bar-right')).toBeInTheDocument();
        });

        it('should apply correct CSS classes for layout', async () => {
            mockUseSession.mockReturnValue({ data: null, status: 'unauthenticated' } as any);
            
            render(await NavBar());
            
            const nav = screen.getByRole('navigation');
            expect(nav).toHaveClass('border-b', 'border-gray-100');
            
            const container = nav.querySelector('div');
            expect(container).toHaveClass('px-4', 'md:px-6', 'lg:px-8');
        });

        it('should pass viewport width to NavBarDesktopRight', async () => {
            mockUseSession.mockReturnValue({ data: null, status: 'unauthenticated' } as any);
            mockGetViewportWidth.mockResolvedValue(768);
            
            render(await NavBar());
            
            expect(screen.getByTestId('nav-bar-desktop-right')).toBeInTheDocument();
        });
    });

    describe('Authentication Context', () => {
        describe('When user is logged out', () => {
            beforeEach(() => {
                mockUseSession.mockReturnValue({ 
                    data: null, 
                    status: 'unauthenticated',
                    update: jest.fn()
                } as any);
            });

            it('should render all navigation components', async () => {
                render(await NavBar());
                
                expect(screen.getByTestId('nav-bar-left')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-desktop-middle')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-desktop-right')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-right')).toBeInTheDocument();
            });

            it('should pass viewport width to NavBarDesktopRight', async () => {
                mockGetViewportWidth.mockResolvedValue(1280);
                
                render(await NavBar());
                
                expect(screen.getByTestId('nav-bar-desktop-right')).toBeInTheDocument();
            });
        });

        describe('When user is logged in', () => {
            beforeEach(() => {
                mockUseSession.mockReturnValue({
                    data: mockSession,
                    status: 'authenticated',
                    update: jest.fn()
                } as any);
            });

            it('should render all navigation components when authenticated', async () => {
                render(await NavBar());
                
                expect(screen.getByTestId('nav-bar-left')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-desktop-middle')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-desktop-right')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-right')).toBeInTheDocument();
            });

            it('should pass viewport width to NavBarDesktopRight when authenticated', async () => {
                mockGetViewportWidth.mockResolvedValue(768);
                
                render(await NavBar());
                
                expect(screen.getByTestId('nav-bar-desktop-right')).toBeInTheDocument();
            });
        });

        describe('When session is loading', () => {
            beforeEach(() => {
                mockUseSession.mockReturnValue({
                    data: null,
                    status: 'loading',
                    update: jest.fn()
                } as any);
            });

            it('should render without crashing during loading state', async () => {
                render(await NavBar());
                
                expect(screen.getByRole('navigation')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-left')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-desktop-middle')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-desktop-right')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-right')).toBeInTheDocument();
            });
        });
    });

    describe('Responsive Behavior', () => {
        it('should handle different viewport widths', async () => {
            mockUseSession.mockReturnValue({ data: mockSession, status: 'authenticated' } as any);
            
            const viewportWidths = [320, 768, 1024, 1280];
            
            for (const width of viewportWidths) {
                mockGetViewportWidth.mockResolvedValue(width);
                
                const { unmount } = render(await NavBar());
                
                const desktopRight = screen.getByTestId('nav-bar-desktop-right');
                expect(desktopRight).toBeInTheDocument();
                
                unmount();
            }
        });

        it('should maintain responsive classes regardless of auth state', async () => {
            const authStates = [
                { data: null, status: 'unauthenticated' },
                { data: mockSession, status: 'authenticated' }
            ];
            
            for (const authState of authStates) {
                mockUseSession.mockReturnValue(authState as any);
                
                const { unmount } = render(await NavBar());
                
                const nav = screen.getByRole('navigation');
                expect(nav).toHaveClass('border-b', 'border-gray-100');
                
                const flexContainer = nav.querySelector('.relative.flex.h-15');
                expect(flexContainer).toHaveClass('items-center', 'justify-between');
                
                unmount();
            }
        });
    });

    describe('Navigation State Management', () => {
        it('should handle different pathname states', async () => {
            const paths = ['/', '/properties', '/properties/add', '/profile'];
            
            mockUseSession.mockReturnValue({ data: mockSession, status: 'authenticated' } as any);
            
            for (const path of paths) {
                mockUsePathname.mockReturnValue(path);
                
                const { unmount } = render(await NavBar());
                
                // Components should receive the correct pathname context
                expect(screen.getByTestId('nav-bar-desktop-middle')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-right')).toBeInTheDocument();
                
                unmount();
            }
        });
    });

    describe('Error Handling', () => {
        it('should handle viewport width fetch failure gracefully', async () => {
            mockUseSession.mockReturnValue({ data: null, status: 'unauthenticated' } as any);
            mockGetViewportWidth.mockRejectedValue(new Error('Failed to fetch viewport width'));
            
            // Should not throw an error
            const renderPromise = NavBar();
            await expect(renderPromise).rejects.toThrow();
        });

        it('should handle missing session data gracefully', async () => {
            mockUseSession.mockReturnValue({ data: undefined, status: 'unauthenticated' } as any);
            
            render(await NavBar());
            
            // Should render all components without crashing
            expect(screen.getByTestId('nav-bar-left')).toBeInTheDocument();
            expect(screen.getByTestId('nav-bar-desktop-middle')).toBeInTheDocument();
            expect(screen.getByTestId('nav-bar-desktop-right')).toBeInTheDocument();
            expect(screen.getByTestId('nav-bar-right')).toBeInTheDocument();
        });
    });

    describe('Accessibility', () => {
        it('should have proper navigation landmark', async () => {
            mockUseSession.mockReturnValue({ data: null, status: 'unauthenticated' } as any);
            
            render(await NavBar());
            
            const nav = screen.getByRole('navigation');
            expect(nav).toBeInTheDocument();
        });

        it('should maintain all navigation components across auth states', async () => {
            const authStates = [
                { data: null, status: 'unauthenticated' },
                { data: mockSession, status: 'authenticated' }
            ];
            
            for (const authState of authStates) {
                mockUseSession.mockReturnValue(authState as any);
                
                const { unmount } = render(await NavBar());
                
                // All main navigation components should be present
                expect(screen.getByTestId('nav-bar-left')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-desktop-middle')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-desktop-right')).toBeInTheDocument();
                expect(screen.getByTestId('nav-bar-right')).toBeInTheDocument();
                
                unmount();
            }
        });
    });

    describe('Integration', () => {
        it('should pass viewport width to NavBarDesktopRight component', async () => {
            mockUseSession.mockReturnValue({ data: mockSession, status: 'authenticated', update: jest.fn() } as any);
            mockGetViewportWidth.mockResolvedValue(1280);
            
            render(await NavBar());
            
            // Verify NavBarDesktopRight receives viewport width
            const desktopRight = screen.getByTestId('nav-bar-desktop-right');
            expect(desktopRight).toBeInTheDocument();
        });

        it('should render all child components regardless of auth state', async () => {
            mockUseSession.mockReturnValue({ data: null, status: 'unauthenticated', update: jest.fn() } as any);
            
            render(await NavBar());
            
            // All child components should be present
            expect(screen.getByTestId('nav-bar-left')).toBeInTheDocument();
            expect(screen.getByTestId('nav-bar-desktop-middle')).toBeInTheDocument();
            expect(screen.getByTestId('nav-bar-desktop-right')).toBeInTheDocument();
            expect(screen.getByTestId('nav-bar-right')).toBeInTheDocument();
        });
    });

    describe('Snapshots', () => {
        it('should match snapshot when logged out', async () => {
            mockUseSession.mockReturnValue({ data: null, status: 'unauthenticated' } as any);
            
            const navBar = await NavBar();
            expect(navBar).toMatchSnapshot();
        });

        it('should match snapshot when logged in', async () => {
            mockUseSession.mockReturnValue({ data: mockSession, status: 'authenticated' } as any);
            
            const navBar = await NavBar();
            expect(navBar).toMatchSnapshot();
        });

        it('should match snapshot with different viewport widths', async () => {
            mockUseSession.mockReturnValue({ data: mockSession, status: 'authenticated' } as any);
            mockGetViewportWidth.mockResolvedValue(768);
            
            const navBar = await NavBar();
            expect(navBar).toMatchSnapshot();
        });
    });
});