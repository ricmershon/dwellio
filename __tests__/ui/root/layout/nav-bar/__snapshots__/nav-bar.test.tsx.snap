// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`NavBar Snapshots should match snapshot when logged in 1`] = `
{
  "$$typeof": Symbol(react.transitional.element),
  "_owner": null,
  "_store": {},
  "key": null,
  "props": {
    "children": {
      "$$typeof": Symbol(react.transitional.element),
      "_owner": null,
      "_store": {},
      "key": null,
      "props": {
        "children": {
          "$$typeof": Symbol(react.transitional.element),
          "_owner": null,
          "_store": {},
          "key": null,
          "props": {
            "children": [
              {
                "$$typeof": Symbol(react.transitional.element),
                "_owner": null,
                "_store": {},
                "key": null,
                "props": {},
                "type": [Function],
              },
              {
                "$$typeof": Symbol(react.transitional.element),
                "_owner": null,
                "_store": {},
                "key": null,
                "props": {},
                "type": [Function],
              },
              {
                "$$typeof": Symbol(react.transitional.element),
                "_owner": null,
                "_store": {},
                "key": null,
                "props": {
                  "children": [
                    {
                      "$$typeof": Symbol(react.transitional.element),
                      "_owner": null,
                      "_store": {},
                      "key": null,
                      "props": {
                        "viewportWidth": 1024,
                      },
                      "type": [Function],
                    },
                    {
                      "$$typeof": Symbol(react.transitional.element),
                      "_owner": null,
                      "_store": {},
                      "key": null,
                      "props": {},
                      "type": [Function],
                    },
                  ],
                  "className": "flex relative",
                },
                "type": "div",
              },
            ],
            "className": "relative flex h-15 items-center justify-between",
          },
          "type": "div",
        },
        "className": "px-4 md:px-6 lg:px-8",
      },
      "type": "div",
    },
    "className": "border-b border-gray-100",
  },
  "type": "nav",
}
`;

exports[`NavBar Snapshots should match snapshot when logged out 1`] = `
{
  "$$typeof": Symbol(react.transitional.element),
  "_owner": null,
  "_store": {},
  "key": null,
  "props": {
    "children": {
      "$$typeof": Symbol(react.transitional.element),
      "_owner": null,
      "_store": {},
      "key": null,
      "props": {
        "children": {
          "$$typeof": Symbol(react.transitional.element),
          "_owner": null,
          "_store": {},
          "key": null,
          "props": {
            "children": [
              {
                "$$typeof": Symbol(react.transitional.element),
                "_owner": null,
                "_store": {},
                "key": null,
                "props": {},
                "type": [Function],
              },
              {
                "$$typeof": Symbol(react.transitional.element),
                "_owner": null,
                "_store": {},
                "key": null,
                "props": {},
                "type": [Function],
              },
              {
                "$$typeof": Symbol(react.transitional.element),
                "_owner": null,
                "_store": {},
                "key": null,
                "props": {
                  "children": [
                    {
                      "$$typeof": Symbol(react.transitional.element),
                      "_owner": null,
                      "_store": {},
                      "key": null,
                      "props": {
                        "viewportWidth": 1024,
                      },
                      "type": [Function],
                    },
                    {
                      "$$typeof": Symbol(react.transitional.element),
                      "_owner": null,
                      "_store": {},
                      "key": null,
                      "props": {},
                      "type": [Function],
                    },
                  ],
                  "className": "flex relative",
                },
                "type": "div",
              },
            ],
            "className": "relative flex h-15 items-center justify-between",
          },
          "type": "div",
        },
        "className": "px-4 md:px-6 lg:px-8",
      },
      "type": "div",
    },
    "className": "border-b border-gray-100",
  },
  "type": "nav",
}
`;

exports[`NavBar Snapshots should match snapshot with different viewport widths 1`] = `
{
  "$$typeof": Symbol(react.transitional.element),
  "_owner": null,
  "_store": {},
  "key": null,
  "props": {
    "children": {
      "$$typeof": Symbol(react.transitional.element),
      "_owner": null,
      "_store": {},
      "key": null,
      "props": {
        "children": {
          "$$typeof": Symbol(react.transitional.element),
          "_owner": null,
          "_store": {},
          "key": null,
          "props": {
            "children": [
              {
                "$$typeof": Symbol(react.transitional.element),
                "_owner": null,
                "_store": {},
                "key": null,
                "props": {},
                "type": [Function],
              },
              {
                "$$typeof": Symbol(react.transitional.element),
                "_owner": null,
                "_store": {},
                "key": null,
                "props": {},
                "type": [Function],
              },
              {
                "$$typeof": Symbol(react.transitional.element),
                "_owner": null,
                "_store": {},
                "key": null,
                "props": {
                  "children": [
                    {
                      "$$typeof": Symbol(react.transitional.element),
                      "_owner": null,
                      "_store": {},
                      "key": null,
                      "props": {
                        "viewportWidth": 768,
                      },
                      "type": [Function],
                    },
                    {
                      "$$typeof": Symbol(react.transitional.element),
                      "_owner": null,
                      "_store": {},
                      "key": null,
                      "props": {},
                      "type": [Function],
                    },
                  ],
                  "className": "flex relative",
                },
                "type": "div",
              },
            ],
            "className": "relative flex h-15 items-center justify-between",
          },
          "type": "div",
        },
        "className": "px-4 md:px-6 lg:px-8",
      },
      "type": "div",
    },
    "className": "border-b border-gray-100",
  },
  "type": "nav",
}
`;
